File: src/2022/Day3/solution.ts
import fs from "fs/promises";

async function main(file: string) {
  const data: string = await fs.readFile(file, { encoding: "utf-8" });
  const rucksacks: Array<string> = data.split("\r\n");

  const itemPriorities = rucksacks.map((rucksack) => {
    const mid = rucksack.length / 2;
    const repeatedItem = getRepeatedItem(rucksack.slice(0, mid), rucksack.slice(mid));
    return determinePriority(repeatedItem!);
  });

  const groups: string[][] = createGroups(rucksacks);
  const badges: Array<string> = groups.map((group) => getGroupBadge(group)!);
  const badgePriorities: Array<number> = badges.map((badge) => determinePriority(badge));

  return {
    PartA: sumPriorities(itemPriorities),
    PartB: sumPriorities(badgePriorities),
  };
}

function getRepeatedItem(firstCompartment: string, secondCompartment: string): string | null {
  for (const char of firstCompartment) {
    if (secondCompartment.includes(char)) {
      return char;
    }
  }
  return null;
}

function determinePriority(char: string): number {
  if (char === char.toLowerCase()) {
    return char.charCodeAt(0) - 96;
  }
  return char.charCodeAt(0) - 38;
}

function sumPriorities(priorities: Array<number>): number {
  return priorities.reduce((total, priority) => total + priority);
}

function createGroups(rucksacks: Array<string>): string[][] {
  const groupSize: number = 3;
  const numOfGroups: number = rucksacks.length / groupSize;
  return Array.from({ length: numOfGroups }, (_, i) => rucksacks.slice(i * groupSize, (i + 1) * groupSize))
}

function getGroupBadge(group: Array<string>): string | null {
  const [sack1, sack2, sack3] = group;
  for (const item of sack1) {
    if (sack2.includes(item) && sack3.includes(item)) {
      return item;
    }
  }
  return null
}

main("input.txt").then((res) => console.log(res));

File: src/2022/Day3/solution.py
with open('input.txt') as f:
    data = f.read()

def determine_priority(char: str) -> int:
    if char == char.lower():
        return ord(char) - 96
    return ord(char) - 38

def find_repeated_item(group: list) -> str:
    item = set(group[0])
    for rucksack in group[1:]:
        item = item.intersection(set(rucksack))
    return item.pop()

rucksacks = data.split("\n")
elf_groups = [rucksacks[i : i + 3] for i in range(0, len(rucksacks), 3)]

item_priorities = []
for rucksack in rucksacks:
    mid = len(rucksack) // 2
    compartments = [rucksack[:mid], rucksack[mid:]]

    item = find_repeated_item(compartments)
    item_priorities.append(determine_priority(item))

badge_priorities = []
for group in elf_groups:
    badge = find_repeated_item(group)
    badge_priorities.append(determine_priority(badge))

print("Part A: ", sum(item_priorities))
print("Part B: ", sum(badge_priorities))
